### ISSUE TITLE
Possible bugs in the ReceiverLink SetCredit method

### ISSUE THREAD

**Issue by billpoole-mi** [Azure/amqpnetlite#326]
I think I may have found a couple of bugs in the ReceiverLink SetCredit method.

this.autoRestore is set on line 155 of ReceiverLink.cs, but if the credit parameter <= this.totalCredit, then the method returns without setting this.autoRestore. That is, this.autoRestore is only set if the credit parameter > this.totalCredit.

Also, if the credit parameter > this.totalCredit, this.restored is set to zero on line 157 but the "restored" count is not added to the remaining credit (this.credit), nor included in the credit sent in the Flow frame. Consequently, I think it is possible for the receiver link to become permanently starved.

For example, if this.totalCredit = 200, this.credit = 101 and this.restored = 99 (and this.pending = 0) and SetCredit is invoked with credit = 202, then this.totalCredit becomes 202, this.credit becomes 103 and this.restored becomes zero. The threshold to auto-restore is now 202 / 2 = 101.

100 messages are then received and processed such that this.credit = 3 and this.restored = 100 (and this.pending still zero). SetCredit is then invoked with credit = 204. Then this.totalCredit becomes 204, this.credit becomes 5 and this.restored becomes zero. The threshold to auto-restore is now 204 / 2 = 102.

However, with this.credit = 5 and this.pending = 0, there is no way for this.restored to ever grow beyond 5 and thus will never reach 102. The receiver link is now permanently starved.

Therefore, I propose that line 136 be updated to include the restored credit in the update the remaining credit and the credit sent in the Flow frame:

```csharp
this.credit += credit - this.totalCredit + this.restored;
```

I believe it is better to do this than to leave line 136 as is and not reset this.restored because since we're going to the effort of sending a Flow frame, we might as well include the outstanding restored credits in that request.

I've updated the SetCredit method below to include fixes for the above. I've also included a check for where autoRestore = false to only send a Flow frame if the remaining credit (this.credit) is changing (for efficiency reasons).

```csharp
public void SetCredit(int credit, bool autoRestore = true)
{
	lock (this.ThisLock)
	{
		if (this.IsDetaching)
		{
			return;
		}

		if (this.totalCredit < 0)
		{
			this.totalCredit = 0;
		}

		var sendFlow = false;
		if (autoRestore)
		{
			// Only change remaining credit if total credit was increased, to allow
			// accepting incoming messages. If total credit is reduced, only update 
			// total so credit will be later auto-restored to the new limit.
			if (credit > this.totalCredit)
			{
				this.credit += credit - this.totalCredit + this.restored;
				this.restored = 0;
				sendFlow = true;
			}
		}
		else
		{
			this.pending = 0;
			this.restored = 0;
			if (credit != this.credit)
			{
				this.credit = credit;
				sendFlow = true;
			}
		}

		this.totalCredit = credit;
		this.autoRestore = autoRestore;
		if (sendFlow)
		{
			this.SendFlow(this.deliveryCount, (uint)this.credit, false);
		}
	}
}
```

**Comment by xinchen10** [comment]
I agree with you on the analysis and the changes seem good to me. Can you create a PR with the fix? The original intention of this API is to allow application to overwrite the default link credit after the receiver link is created, or completely stop the link by setting credit to 0. I haven't tested it thoroughly with scenarios like yours where credit is changing frequently. I will try to create a unit test that simulates this scenario. Thanks a lot for looking into this.

**Comment by billpoole-mi** [comment]
You're welcome! The pull request has been created.

**Comment by Necriis** [comment]
@xinchen10 Do you plan a release soon regarding this bug ?
That is blocking for my app...

**Comment by xinchen10** [comment]
Released 2.1.7
